File: GGAndroid_2.0/GGDemo/src/main/java/io/ganguo/demo/viewmodel/activity/CoordinatorAppBarDemoVModel.java
package io.ganguo.demo.viewmodel.activity;

import android.support.design.widget.AppBarLayout;
import android.view.View;
import android.view.ViewGroup;


import java.util.Arrays;

import io.ganguo.demo.R;
import io.ganguo.demo.viewmodel.item.ItemBannerDemoVModel;
import io.ganguo.demo.viewmodel.item.ItemTabVModel;
import io.ganguo.demo.viewmodel.page.TabPageViewModel;
import io.ganguo.vmodel.BaseViewModel;
import io.ganguo.vmodel.ViewModelHelper;
import io.ganguo.library.ui.view.ActivityInterface;
import io.ganguo.utils.util.Colors;
import io.ganguo.viewmodel.common.CommonViewPagerVModel;
import io.ganguo.viewmodel.common.CoordinatorAppBarVModel;
import io.ganguo.viewmodel.common.HeaderItemViewModel;
import io.ganguo.viewmodel.common.HeaderViewModel;
import io.ganguo.viewmodel.common.TabLayoutViewModel;
import io.ganguo.viewmodel.databinding.IncludeCoordinatorLayoutAppBarBinding;

/**
 * <p>
 * 1、CoordinatorAppBarVModel Demo
 * 2、通过CoordinatorLayout和AppBarLayout可以实现原生折叠/计算颜色渐变效果
 * <p/>
 * Created by leo on 2018/7/13.
 */
public class CoordinatorAppBarDemoVModel extends CoordinatorAppBarVModel<ActivityInterface<IncludeCoordinatorLayoutAppBarBinding>> {
    private CommonViewPagerVModel viewPagerVModel;
    private ItemBannerDemoVModel bannerViewModel;

    @Override
    public void onViewAttached(View view) {
        hideLoading();
    }


    @Override
    public void initAppBarHeader(ViewGroup container) {
        super.initAppBarHeader(container);
        ViewModelHelper.bind(container, this, bannerViewModel = new ItemBannerDemoVModel());
    }

    /**
     * function: init AppBarContent
     *
     * @param container
     */
    @Override
    public void initAppBarContent(ViewGroup container) {
        super.initAppBarContent(container);
        ViewModelHelper.bind(container, this, onCreateTabViewModel());
    }

    /**
     * function: init Content
     *
     * @param container
     */
    @Override
    public void initContent(ViewGroup container) {
        super.initContent(container);
        ViewModelHelper.bind(container, this, getViewPagerVModel());
    }

    @Override
    public void onOffsetChanged(AppBarLayout appBarLayout, int verticalOffset) {
        super.onOffsetChanged(appBarLayout, verticalOffset);
        //可根据滑动的距离，在此处计算动态设置一些View颜色/透明度
        float factor = Math.abs((float) verticalOffset / appBarLayout.getTotalScrollRange());
        int color = Colors.evaluatorViewColor(factor, getColors(R.color.transparent), getColors(R.color.blue_dark));
        bannerViewModel.getRootView().setBackgroundColor(color);
    }


    /**
     * function：create TabLayoutViewModel
     *
     * @return
     */
    protected TabLayoutViewModel onCreateTabViewModel() {
        TabLayoutViewModel tabViewVModel = new TabLayoutViewModel.Builder(this)
                .indicatorColor(getColors(R.color.colorAccent), getColors(R.color.colorAccent))
                .indicatorWidth(getDimensionPixelOffsets(R.dimen.dp_30))
                .indicatorHeight(getDimensionPixelOffsets(R.dimen.tab_height))
                .distributeEvenly(true)
                .indicatorVisible(true)
                .setIndicatorWidthWrapContent(false)
                .setViewPagerSmoothScroll(true)
                .setViewPagerSmoothScrollAnimation(true)
                .setBackgroundColor(getColors(R.color.gray))
                .setLayoutId(R.layout.include_tab_horizontal_layout_view_model)
                .appendViewModel(new ItemTabVModel(R.drawable.selector_tab_icon, "tab1"))
                .appendViewModel(new ItemTabVModel(R.drawable.selector_tab_icon, "tab2"))
                .appendViewModel(new ItemTabVModel(R.drawable.selector_tab_icon, "tab3"))
                .bindControlScrollViewPager(getViewPagerVModel().getViewPager())
                .build();
        return tabViewVModel;
    }


    /**
     * function：create CommonViewPagerVModel
     *
     * @return
     */
    public CommonViewPagerVModel getViewPagerVModel() {
        if (viewPagerVModel == null) {
            viewPagerVModel = new CommonViewPagerVModel(
                    Arrays.<BaseViewModel>asList(
                            new TabPageViewModel(),
                            new TabPageViewModel(),
                            new TabPageViewModel()
                    )
            );

        }
        return viewPagerVModel;
    }
}

File: GGAndroid_3.0/main/app-mvvm-v1-sample/src/main/java/io/ganguo/demo/viewmodel/activity/CoordinatorAppBarDemoVModel.kt
package io.ganguo.demo.viewmodel.activity

import android.view.View
import android.view.ViewGroup
import com.scwang.smart.refresh.layout.api.RefreshLayout
import io.ganguo.demo.R
import io.ganguo.demo.viewmodel.item.ItemBannerDemoVModel
import io.ganguo.demo.viewmodel.item.ItemTabVModel
import io.ganguo.demo.viewmodel.page.TabPageViewModel
import io.ganguo.viewmodel.core.viewinterface.ActivityInterface
import io.ganguo.state.loading.ILoadingHandler
import io.ganguo.state.loading.LoadingHelper
import io.ganguo.utils.util.Colors
import io.ganguo.utils.util.postDelayed
import io.ganguo.viewmodel.pack.common.CommonViewPagerVModel
import io.ganguo.viewmodel.pack.common.CoordinatorAppBarVModel
import io.ganguo.viewmodel.pack.common.TabLayoutViewModel
import io.ganguo.viewmodel.pack.common.loading.LoadingMaterialProgressVModel
import io.ganguo.viewmodel.databinding.IncludeCoordinatorLayoutAppBarBinding
import io.ganguo.viewmodel.pack.helper.LoadingVModelHelper
import io.ganguo.viewmodel.core.BaseViewModel
import io.ganguo.viewmodel.core.ViewModelHelper

/**
 *
 *
 * 1、CoordinatorAppBarVModel Demo
 * 2、通过CoordinatorLayout和AppBarLayout可以实现原生折叠/计算颜色渐变效果
 *
 *
 * Created by leo on 2018/7/13.
 */
open class CoordinatorAppBarDemoVModel : CoordinatorAppBarVModel<ActivityInterface<IncludeCoordinatorLayoutAppBarBinding>>(), ILoadingHandler {
    override val loadingHelper: LoadingHelper by lazy {
        LoadingVModelHelper.bindLoadingView(loadingContainer, this, LoadingMaterialProgressVModel())
    }

    private lateinit var bannerViewModel: ItemBannerDemoVModel
    private val viewPagerVModel: CommonViewPagerVModel by lazy {
        newViewPagerVModel()
    }


    override fun onViewAttached(view: View) {
        enableAppBarElevation()
        showLoading()
        context.postDelayed(1000) { hideLoading() }
    }


    override fun initAppBarHeader(container: ViewGroup) {
        super.initAppBarHeader(container)
        bannerViewModel = ItemBannerDemoVModel()
        ViewModelHelper.bind(container, this, bannerViewModel)
    }

    /**
     * function: init AppBarContent
     *
     * @param container
     */
    override fun initAppBarContent(container: ViewGroup) {
        super.initAppBarContent(container)
        ViewModelHelper.bind(container, this, newTabViewModel())
    }

    /**
     * function: init Content
     *
     * @param container
     */
    override fun initContent(container: ViewGroup) {
        super.initContent(container)
        ViewModelHelper.bind(container, this, viewPagerVModel)
    }


    override fun onAppBarLayoutExpandState(isExpand: Boolean, factor: Float) {
        super.onAppBarLayoutExpandState(isExpand, factor)
        //可根据滑动的距离，在此处计算动态设置一些View颜色/透明度
        val color = Colors.evaluatorViewColor(factor, getColor(R.color.transparent), getColor(R.color.blue_dark))
        bannerViewModel.rootView.setBackgroundColor(color)
    }


    /**
     * 生成 TabLayoutViewModel
     *
     * @return
     */
    private fun newTabViewModel(): TabLayoutViewModel {
        return TabLayoutViewModel.Builder(this)
                .indicatorColor(getColor(R.color.colorAccent), getColor(R.color.colorAccent))
                .indicatorWidth(getDimensionPixelOffset(R.dimen.dp_30))
                .indicatorHeight(getDimensionPixelOffset(R.dimen.tab_height))
                .distributeEvenly(true)
                .indicatorVisible(true)
                .indicatorWidthWrapContent(false)
                .viewPagerSmoothScroll(true)
                .viewPagerSmoothScrollAnimation(true)
                .backgroundColor(getColor(R.color.white))
                .layoutId(R.layout.include_tab_horizontal_layout_view_model)
                .appendViewModel(ItemTabVModel(R.drawable.selector_tab_icon, "tab1"))
                .appendViewModel(ItemTabVModel(R.drawable.selector_tab_icon, "tab2"))
                .appendViewModel(ItemTabVModel(R.drawable.selector_tab_icon, "tab3"))
                .bindControlScrollViewPager(viewPagerVModel.viewPager)
                .build()
    }

    /**
     * 生成 CommonViewPagerVModel
     *
     * @return
     */
    private fun newViewPagerVModel(): CommonViewPagerVModel {
        return CommonViewPagerVModel(
                listOf<BaseViewModel<*>>(
                        TabPageViewModel(),
                        TabPageViewModel(),
                        TabPageViewModel()
                )
        )
    }

    override fun onRefresh(refreshLayout: RefreshLayout) {
        super.onRefresh(refreshLayout)
        refreshLayout.finishRefresh()
    }

}
